components/feature/webnotifications/src/main/java/mozilla/components/feature/webnotifications/WebNotificationFeature.kt

```diff
@@ -0,0 +1,104 @@
+import mozilla.components.concept.engine.webnotifications.WebNotificationDelegate
+import mozilla.components.support.base.log.logger.Logger
+import java.lang.UnsupportedOperationException
+
+private const val NOTIFICATION_GROUP_ID = "mozac.feature.webnotifications.generic.group"
+private const val NOTIFICATION_CHANNEL_ID = "mozac.feature.webnotifications.generic.channel"
+
+/**
+ * Feature implementation for configuring and displaying web notifications to the user.
+ *
+ * Initialize this feature globally once on app start
+ * ```Kotlin
+ * WebNotificationFeature(
+ *     applicationContext, engine, icons, R.mipmap.ic_launcher, BrowserActivity::class.java
+ * )
+ * ```
+ *
+ * @param context The application Context.
+ * @param engine The browser engine.
+ * @param browserIcons The entry point for loading the large icon for the notification.
+ * @param smallIcon The small icon for the notification.
+ * @param activityClass The Activity that the notification will launch if user taps on it
+ */
+class WebNotificationFeature(
+    private val context: Context,
+    private val engine: Engine,
+    private val browserIcons: BrowserIcons,
+    @DrawableRes private val smallIcon: Int,
+    private val activityClass: Class<*>?
+) : WebNotificationDelegate {
+    private val logger = Logger("AutoPushFeature")
+    private var pendingRequestId = 0
+    private var notificationId = 0
+    private val notificationIdMap = HashMap<String, Int>()
+    private val notificationManager = context.getSystemService<NotificationManager>()
+    private val nativeNotificationBridge = NativeNotificationBridge(browserIcons, smallIcon)
+
+    init {
+        try {
+            engine.registerWebNotificationDelegate(this)
+        } catch (e: UnsupportedOperationException) {
+            logger.error("failed to register for web notification delegate", e)
+        }
+    }
+
+    override fun onShowNotification(webNotification: WebNotification) {
+        ensureNotificationGroupAndChannelExists()
+
+        notificationIdMap[webNotification.tag]?.let {
+            notificationManager?.cancel(it)
+        }
+
+        notificationIdMap[webNotification.tag] = notificationId
+        GlobalScope.launch(Dispatchers.IO) {
+            notificationManager?.notify(notificationId, nativeNotificationBridge.convertToAndroidNotification(
+                webNotification, context, NOTIFICATION_CHANNEL_ID, activityClass, pendingRequestId))
+            pendingRequestId++
+            notificationId++
+        }
+    }
+
+    override fun onCloseNotification(webNotification: WebNotification) {
+        notificationIdMap[webNotification.tag]?.let {
+            notificationManager?.cancel(it)
+        }
+    }
+
+    private fun ensureNotificationGroupAndChannelExists() {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            notificationManager?.createNotificationChannelGroup(NotificationChannelGroup(
```

Comment:
Mentioned above as well, we probably want the notification settings to be in site permissions, so we don't need a group here. We can put it at the top level like Send Tab does.

Reply:
Ah, I'll remove the group.